text/0035-schema-applies-to-unspecified.md

```diff
@@ -0,0 +1,62 @@
+
+## Related issues and PRs
+
+- Reference Issues: [cedar#351](https://github.com/cedar-policy/cedar/issues/351)
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2023-10-25
+
+## Summary
+
+This RFC proposes to change the semantics of the `appliesTo` field in the schema, as summarized by the following table.
+| Scenario | Status Quo | Proposal |
+| --------- | -------- | ------- |
+| (1) Specified `appliesTo` but omitted `principalTypes`/`resourceTypes` | Authz request can have a principal/resource entity of any type, or the unspecified entity | Authz request with this action can only have unspecified entity as principal/resource |
+| (2) Omitted `appliesTo` | Authz request can have a principal/resource entity of any type, or the unspecified entity | Action cannot be used in authz request |
+| (3) `appliesTo.principalTypes: []` / `appliesTo.resourceTypes: []` | Action cannot be used in authz request | ❌ Not allowed |
+| (4) `appliesTo` is defined but all sub-elements are omitted | Authz request can have a principal/resource entity of any type, or the unspecified entity; context is an empty record | ❌ Not allowed |
+
+## Motivation
+
+**Current behavior**  
+In Cedar 2.x, if an action's `appliesTo.principalTypes` or `appliesTo.resourceTypes` is not given (or if the entire `appliesTo` element is not given), then it's interpreted as an action that applies to all principal types and resource types.
+
+**Challenges**  
+Discussion on [RFC 24](https://github.com/cedar-policy/rfcs/pull/24) highlighted challenges this introduces, primarily the risk of unintentionally specifying that an action applies to all principal types / resource types and related complications it causes for analysis and the experience of policy validation as error messages become more confusing.
+
+**Feature request**  
+We can mitigate the challenges listed above by making following changes:
+
+1. Specified `appliesTo` but omitted `appliesTo.principalTypes` / `appliesTo.resourceTypes` means that the request component is unspecified, i.e., corresponding to the `None` option in the principal and/or resource component of a [`Request`](https://docs.rs/cedar-policy/2.4.0/cedar_policy/struct.Request.html). Omitted `appliesTo.context` means that the context is an empty record (same as the status quo).
+2. Omitted `appliesTo` means that the action cannot be used in a request. It is used exclusively as an action group.
+3. Disallow empty arrays for `appliesTo.principalTypes` / `appliesTo.resourceTypes`.
+4. Disallow an empty `appliesTo` attribute thus requiring at least one of `principalTypes`, `resourceTypes` or `context` to be specified if `appliesTo` is specified.
+
+This feature request is consistent with the semantics for `appliesTo` proposed in [RFC 24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md) for the alternate Cedar schema syntax.
+
+## Detailed design
+
+This RFC will require the following changes:
```

Comment:
This RFC should also consider the semantics of `null` values. The schema parser currently represents this as `Option<Vec<_>>` which `serde_json` handles by deserializing both a `null` or an omitted value as `None`. This might not be what we want.

Reply:
I don't think `null` values belong here, at least not at this late date as things are winding down. Let's make a separate RFC for that.

Reply:
The RFC was posted a week ago and isn't in FCP? Does not seem too late to add this

Reply:
Not too late 🙂 but I think @mwhicks1's point was that we want a separate RFC to decide on uniform treatment for `null` across all Cedar JSON representations, rather than just this RFC.

Personally, I think it makes sense to treat `null` values the same way as an omitted value.

Reply:
>  it makes sense to treat null values the same way as an omitted value.

Agree. Let's put that it in its own RFC. At this rate, it will get approved faster than this one!

Reply:
I'm going to resolve this thread because I think it would be better addressed in #37. Will leave comments there soon.